src/elliptic/curves/wrappers.rs

```diff
@@ -0,0 +1,1692 @@
+        let scalar = self.as_raw().add(scalar.as_raw());
+        Self::from_raw(scalar)
+    }
+
+    /// Subtracts two scalars, returns the result by modulo `q`, or `None` if resulting scalar is zero
+    pub fn sub_checked(&self, scalar: &Scalar<E>) -> Result<Self, ZeroScalarError> {
+        let scalar = self.as_raw().sub(scalar.as_raw());
+        Self::from_raw(scalar)
+    }
+
+    /// Multiplies two scalars, returns the result by modulo `q`, or `None` if resulting scalar is zero
+    pub fn mul_checked(&self, scalar: &Scalar<E>) -> Result<Self, ZeroScalarError> {
+        let scalar = self.as_raw().mul(scalar.as_raw());
+        Self::from_raw(scalar)
+    }
+
+    fn from_raw(raw_scalar: E::Scalar) -> Result<Self, ZeroScalarError> {
+        if raw_scalar.is_zero() {
+            Err(ZeroScalarError(()))
+        } else {
+            Ok(Self { raw_scalar })
+        }
+    }
+
+    unsafe fn from_raw_unchecked(raw_scalar: E::Scalar) -> Self {
+        Self { raw_scalar }
+    }
+
+    fn as_raw(&self) -> &E::Scalar {
+        &self.raw_scalar
+    }
+
+    fn into_raw(self) -> E::Scalar {
+        self.raw_scalar
+    }
+}
+
+impl<E: Curve> Clone for Scalar<E> {
+    fn clone(&self) -> Self {
+        // Safety: `self` is guaranteed to be non-zero
+        unsafe { Scalar::from_raw_unchecked(self.as_raw().clone()) }
+    }
+}
+
+impl<E: Curve> fmt::Debug for Scalar<E> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.as_raw().fmt(f)
+    }
+}
+
+impl<E: Curve> PartialEq for Scalar<E> {
+    fn eq(&self, other: &Self) -> bool {
+        self.as_raw().eq(other.as_raw())
+    }
+}
+
+impl<E: Curve> PartialEq<ScalarZ<E>> for Scalar<E> {
+    fn eq(&self, other: &ScalarZ<E>) -> bool {
+        self.as_raw().eq(other.as_raw())
+    }
+}
+
+impl<E: Curve> TryFrom<u16> for Scalar<E> {
+    type Error = ZeroScalarError;
+    fn try_from(n: u16) -> Result<Self, ZeroScalarError> {
+        Self::from_bigint(&BigInt::from(n))
+    }
+}
+
+impl<E: Curve> TryFrom<u32> for Scalar<E> {
+    type Error = ZeroScalarError;
+    fn try_from(n: u32) -> Result<Self, ZeroScalarError> {
+        Self::from_bigint(&BigInt::from(n))
+    }
+}
+
+impl<E: Curve> TryFrom<u64> for Scalar<E> {
+    type Error = ZeroScalarError;
+    fn try_from(n: u64) -> Result<Self, ZeroScalarError> {
+        Self::from_bigint(&BigInt::from(n))
+    }
+}
+
+impl<E: Curve> TryFrom<i32> for Scalar<E> {
+    type Error = ZeroScalarError;
+    fn try_from(n: i32) -> Result<Self, ZeroScalarError> {
+        Self::from_bigint(&BigInt::from(n))
+    }
+}
+
+impl<E: Curve> TryFrom<&BigInt> for Scalar<E> {
+    type Error = ZeroScalarError;
+    fn try_from(n: &BigInt) -> Result<Self, ZeroScalarError> {
+        Self::from_bigint(n)
+    }
+}
+
+impl<E: Curve> TryFrom<BigInt> for Scalar<E> {
+    type Error = ZeroScalarError;
+    fn try_from(n: BigInt) -> Result<Self, ZeroScalarError> {
+        Self::from_bigint(&n)
+    }
+}
+
+macro_rules! matrix {
```

Comment:
Oh god

Reply:
So Mul/Add etc *are* implemented but via the macro

Reply:
You can probably uplift all the reference variations of the implementation into the macro, so it will implement it for the references too

Reply:
I had the same (I think) idea of uplifting the reference, but things got more difficult when I introduced `PointRef`. We should not implement operators for `&PointRef`, as it's the same as implementing operators for `&&Point` (we can do this, but I don't like it).

If you have any ideas how to simplify the macro - feel free to give it a try, and suggest your changes!